### ISSUE TITLE
Apply augmentations on Audio and Spectrogram data

### ISSUE THREAD

**Issue by mogwai** [pyannote/pyannote-audio#498]
Discuss and decide on where augmentations should be applied on data incoming to the model. 

---
Current implementation is to:

- Register augmentations on the input or the output of a torch Module

Potential Ideas:

- Implementing a list of augmentations within the task `task.train__iter__` @hbredin 
- Dedicated class to manage the pipeline of transformations required for a task.   

---

There are two types of data to apply augmentations too, spectrograms and waveforms.
 
In general I think we can assume that augmentations are going to happen before becoming the input of a model. 

```
IO -> Augment -> Model
```

fastai has quite a good api. It allows you to write augmentations that can be applied to different types of data. So for us this could be audio and spectrograms. I'm not sure how useful that is though. https://docs.fast.ai/tutorial.albumentations

- Per Item Augmentations (Not sure how many of these there are if we are using torchaudiomentations)
- Batch Applicable Augmentations (MFCC) 
- Cache Augmentations (Resample)
- Random Augmentations (Spec Augment)

We could create an indexable class, AugmentationPipeline, that manages all of these and allows for a user to add their own augmentations easily as well. The benefit of this would be that you could verify you augmentations on data without needing to train. The class would be an extension of the [pytorch transforms](https://pytorch.org/docs/stable/torchvision/transforms.html) design, adding flexibility depending on input data.

It makes the most sense to assign a augmentation pipeline to the task so it can be like you suggested, used in the train.__iter__ method. However the model itself may be important in deciding what Augmentation Pipeline to use. For example a model that doesn't need spectrograms as it works on raw audio could turn off spectrogram transforms in the Pipeline. 

This pipeline could ask be configured for different tasks which may require different augmentations.

```py
transforms = AugmentationPipeline([RemoveSilence, MFCC, SpecAugment]) 
audio_loader = Audio()
waveform, sr = audio_loader(file)
out = transforms[:1](waveform)
to_pil_image(out)
```

**Comment by hbredin** [comment]
Most of what you suggest is the whole point of [`torch_audiomentations`](https://github.com/asteroid-team/torch-audiomentations/) (itself inspired by audiomentations, itself inspired by albumentations that you mention) to which I plan to keep contributing. So let's make `torch_audiomentations` a first-class citizen of pyannote.audio.

I don't necessarily agree that augmentation can only be applied to audio and spectrogram. 
If we assume that every model should take audio as input and that non-trainable feature extraction such as spectogram is done on GPU as part of the model, then I'd rather distinguish augmentation on the input of a model (audio) vs. augmentation on the inside of the model (e.g. SpecAugment but we could probably think about many others -- Dropout can be applied anywhere and could be considered as an augmentation).

I'd suggest we start by implementing augmentation on the audio input. This would be passed as a `torch_audiomentations` transform (that can be a [composition](https://github.com/asteroid-team/torch-audiomentations/blob/218e3bf61fc87c56ed1ef78dbccee23a1df4fa8a/torch_audiomentations/core/composition.py#L7) of multiple transforms as you suggest) to the `Task`.
The task would then use this transform to augment the training samples yielded by `train__iter__`.

Then, for augmentation on the inside of the model, this could be part of the `Model` itself (either using the existing `register_augmentation` API or a variant, or by defining the augmentation by hand directly in the model).

**Comment by mogwai** [comment]
I'd never thought of Dropout as an augmentation but I guess it is. An augmentation on the Nueral Network itself though?

This sounds good to me. Do you think we should have a method that returns a set of good standard transformations that are known to improve results generally in the voice domain? 

```
Resample(16000) -> MFCC() -> SpecAugment()
```

Returned by `pyannote.audio.core.augmentation.voice_augmentations` ?

I can't think of any augmentations that we want to do that aren't processed on the GPU for now?

**Comment by hbredin** [comment]
We can definitely provide good default augmentation -- but I do think it remains a research problem...

**Comment by mogwai** [comment]
Closed by #609
